qiime/sdk/method.py

```diff
@@ -157,32 +157,13 @@ def _from_markdown(self, markdown_filepath):
         self._init(id_, signature, function, ('markdown', markdown_filepath),
                    name, description, template)
 
-    # TODO this is mostly duplicated from
-    # qiime.core.archiver.Archiver._parse_type. Refactor!
     @classmethod
     def _parse_semantic_type(cls, type_exp):
         # Split the type expression into its components: the semantic_type_exp
-        # and the view_type. Note that this differs from the type definitions
-        # in Archiver._parse_type, as those won't have view types.
+        # and the view_type.
         semantic_type_exp, view_type = type_exp
         view_type = cls._parse_view_type(view_type)
-
-        semantic_type_exp = semantic_type_exp.split('\n')
-        if len(semantic_type_exp) != 1:
-            raise TypeError("Multiple lines in type expression of"
-                            " artifact. Will not load to avoid arbitrary"
-                            " code execution.")
-        semantic_type_exp, = semantic_type_exp
-
-        if ';' in semantic_type_exp:
-            raise TypeError("Invalid type expression in artifact. Will not"
-                            " load to avoid arbitrary code execution.")
-
-        pm = qiime.sdk.PluginManager()
-        locals_ = {k: v[1] for k, v in pm.semantic_types.items()}
-        # Set up all of the types we know about in local scope of the eval
-        # so that complicated type expressions are evaluated.
-        type_ = eval(semantic_type_exp, {'__builtins__': {}}, locals_)
+        type_ = qiime.sdk.parse_type(semantic_type_exp)
```

Comment:
It would also be nice to remove most of the logic in `_parse_primitive_type` so that it just calls `parse_type` and verifies the type is a primitive. `_parse_primitive_type` is currently hardcoded to a few primitives and will break soon.

Thinking about this API more, perhaps it makes sense to have `parse_type` accept an optional parameter that specifies the allowed type expressions? e.g. a parameter that indicates primitive vs semantic vs visualization, and by default allows all defined types.